# feat: 콘서트 조회/포인트 충전/예약-결제 로직 구현 및 단위 테스트 추가

moonyaejin opened the pull request:
## 주요 구현 사항
- **Concert 조회 (레이어드 아키텍처)**
  - ConcertService, ConcertController 구현
  - 공연 목록 및 좌석 조회 기능 추가
  - 단위 테스트: 공연 조회, 특정 ID 조회, 예약된 좌석 제외 후 조회 검증

- **Wallet 충전/결제 (레이어드 아키텍처)**
  - UserWallet 엔티티, WalletJpaAdapter 구현
  - 충전, 결제, 잔액 조회 기능 추가
  - 단위 테스트: 충전/결제 성공, 잔액 부족 예외, 잔액 조회 검증

- **예약/결제 (클린 아키텍처)**
  - SeatHoldPort, HoldSeatUseCase 등 포트/어댑터 분리
  - 좌석 점유/해제, 결제 완료 처리 로직 구현
  - Mock 기반 단위 테스트: 이미 확정된 좌석, 이미 홀드된 좌석, 정상 홀드 성공 검증

---

## 리뷰 받고 싶은 내용
- **커밋 : 2f196ac, e8357dc, cb63bf7**
  예약/결제 기능을 UseCase 단위로 나누고 외부 의존성을 Mock으로 대체했는데, 지금처럼 `SeatHoldPort`, `ReservationRepository` 등을 다 Mock 처리하는 방식이 과연 적절한지 궁금합니다. 혹시 제가 너무 단순하게 분리한 건 아닌지, 충분한지 궁금합니다.

**도메인과 JPA 엔티티 분리 필요성**
- **커밋 : 4878466, b2caee3**
  `Wallet`(도메인)과 `UserWallet`(JPA 엔티티)을 분리했는데, 실제로는 역할이 많이 겹치는 것 같아 중복 코드처럼 보였습니다.  
  과제 수준에서는 그냥 합쳐도 됐을까요? 아니면 장기적으로는 지금처럼 분리하는 게 좋은 선택일까요?  

**중복과 구조적인 실수**  
  구현하면서 `WalletService`와 `WalletJpaAdapter`가 동시에 존재하는 등 중복된 구조가 생겼다가 뒤늦게 정리했습니다.  
  이런 상황을 피하려면 처음 설계할 때 어떤 기준으로 “여긴 Adapter”, “여긴 Service”라고 선을 긋는 게 좋을까요? 

**레이어드 vs 클린 아키텍처 경계**  
- **커밋 : 235dff8, cb63bf7**
  이번 과제에서는 Concert 조회/Wallet 충전은 레이어드로, 예약/결제는 클린 아키텍처로 나눴는데요,  
  실제 현업에서도 이렇게 혼합해서 쓰는 경우가 많은지, 아니면 보통 일관되게 하나로 가져가는지 궁금합니다.  

**Service vs Adapter 경계와 네이밍이 충분히 읽히나요?**
- **커밋 : 279b635**
초기에 책임이 섞였던 걸 *UseCase / *Port / *JpaAdapter / *JpaRepository로 정리했습니다. 지금 구조/네이밍만으로도 역할이 명확히 보이는지, 더 나을 수 있는 기준이 있다면 코멘트 부탁드립니다.
---

## 과제 셀프 피드백
이번 과제는 비즈니스 로직 구현 + 단위 테스트 작성이 목표였는데, 처음에는 테스트를 어디까지 Mock으로 처리해야 하는지가 제일 헷갈렸습니다.  

특히 예약/결제 시나리오는 클린 아키텍처 스타일로 처음 나누어보다 보니, 포트로 뺄 것과 UseCase 안에 둘 것을 구분하는 과정에서 여러 번 리팩터링을 거쳤습니다. 

Concert 조회나 Wallet 충전처럼 단순한 기능은 서비스 + 리포지토리만 두고 가볍게 단위 테스트를 붙였습니다.  그런데 막상 해보니 “이게 진짜 단위 테스트 맞나? 통합 테스트랑 뭐가 다른 거지?”라는 고민도 들었습니다. 그래도 Mock으로 외부 의존성을 끊고 순수한 로직만 검증하는 경험을 해볼 수 있었고, 테스트를 붙여놓으니 작은 로직도 안심하고 리팩터링할 수 있었습니다.  

솔직히 구현하면서 재미도 있었지만, 아직 제가 많이 부족하다는 것도 느꼈습니다.  
DDD를 제대로 지키지 못하고 뒤죽박죽으로 구조를 잡거나, 중복 코드를 만들었던 순간도 많았습니다.  
다음 주차를 시작하기 전에 구조와 부족했던 부분들을 다시 점검하고 넘어가야겠다고 다짐했습니다... 😭

---

## 기술적 성장
- **클린 아키텍처 경험**  
  예약/결제 기능을 유스케이스, 도메인, 포트/어댑터로 나눠서 구현하면서 책임 분리의 장점을 체감했습니다.  

- **Mock 활용**  
  Mock을 어디까지 써야 하는지 처음엔 감이 없었지만,  외부 시스템과 무관하게 빠른 피드백을 얻는 방법을 조금은 배운 것 같습니다.  
  
- **도메인/영속 분리 경험**
  JpaEntity와 도메인 모델을 분리하고 매퍼로 연결해, 기술 의존을 도메인에서 배제하는 방향을 체득했습니다. 앞으로는 도메인 객체에 행위/불변식을 더 밀어 넣어 모델의 응집도를 높일 계획입니다.

moonyaejin committed:
feat: 확정 예약/지갑 DB 스키마 및 JPA 엔티티·리포지토리 초안 추가

moonyaejin committed:
feat: reservation 서비스 뼈대 구현 (유스케이스/어댑터/예외/테스트 등 )

moonyaejin committed:
feat: 좌석 조회를 레이어드로 분리 및 조회 API 추가

moonyaejin committed:
build : Spring Boot 3.4 세팅/의존성 추가 및 JPA/Redis 설정

moonyaejin committed:
db: Flyway 마이그레이션 추가 (reservation, wallet, concert, user)

moonyaejin committed:
feat: 좌석/정책/예외 추가

moonyaejin committed:
feat: Money/LedgerReason/Wallet/IdempotencyKey 도입

moonyaejin committed:
feat: QueueToken/QueuePolicy, User/UserRepository 추가

moonyaejin committed:
feat: ConfirmedReservation/SeatHold/Queue/Wallet 포트 정의

moonyaejin committed:
feat: 엔티티/리포지토리 추가

moonyaejin committed:
feat: JPA/Redis 어댑터 구현

moonyaejin committed:
feat: 좌석 홀드/결제확정 유즈케이스 추가

moonyaejin committed:
feat: 잔액/대기열 유즈케이스 추가

moonyaejin committed:
feat: SeatQueryService 추가

moonyaejin committed:
feat: 응답 DTO 추가

moonyaejin committed:
feat: 응답 DTO 추가

moonyaejin committed:
test: 홀드/결제확정 유즈케이스 단위 테스트 추가

moonyaejin committed:
feat: Concert/ConcertSchedule 모델 및 리포지토리 포트 추가

moonyaejin committed:
feat: ConcertUseCase & ConcertService & ConcertRepository JPA 구현 추가

moonyaejin committed:
feat: InsufficientBalance 예외 추가

moonyaejin committed:
feat: 예약/결제/포인트/대기열 핵심 도메인 및 서비스 로직 구현

moonyaejin committed:
feat: 컨트롤러 및 인프라 어댑터 추가

moonyaejin committed:
feat: JPA 엔티티 및 레포지토리 추가

moonyaejin committed:
test: 서비스 단위 테스트 추가

moonyaejin committed:
feat: 글로벌 예외 핸들러 및 서버 애플리케이션 진입점 수정

moonyaejin committed:
feat: JPA 기반 지갑 엔티티/어댑터 및 단위 테스트 추가

moonyaejin committed:
refactor: port·adapter·JPA 네이밍/위치 정리로 계층 경계 명확화

moonyaejin committed:
refactor(web): Controller를 web 레이어로 이동 및 UseCase 의존으로 통일

moonyaejin committed:
feat(dto): 요청/응답 DTO 도메인별 분리 및 Bean Validation 적용

moonyaejin committed:
refactor: 지갑 유스케이스 단일화 및 음수 금액 가드 추가

moonyaejin committed:
test: 패키지 이동 및 UseCase 모킹으로 테스트 보정

moonyaejin committed:
refactor: 파일 이동

moonyaejin committed:
refactor: Controller를 web 레이어로 이동하고 port-in 의존으로 통일

moonyaejin committed:
refactor: 서비스 구현을 port-in/port-out 구조에 맞게 정리

song-younghoon commented:
헉 PR 분량이 엄청나군요 .. ㅎㅎ
이번 리뷰는 이번 챕터에서 다룬 '아키텍처'에 초점을 맞추어 중점적으로 다뤄보도록 할게요!

한번에 너무 많은 걸 하려 하면 혼란스러우실 수 있으니 이번 주에는 이번 챕터의 토픽에 집중하고,
Persistence와 관련된 세부 구현(JPA, Redis 등) 사항 등은 차후에 관련된 챕터에서 다뤄보도록 합시다 😄

song-younghoon commented on src/main/java/kr/hhplus/be/server/domain/payment/model/Payment.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.time.LocalDateTime;
+
+public class Payment {
```
도메인 객체를 immutable하게 다루고 있군요 👍 
일반적으로 프로젝트를 안전하게 다룰 수 있는 방법입니다.

다만, 함수형 프로그래밍의 패러다임에 익숙치 않은 개발자와 협업한다면 오해가 있을 수 있으니,
README.md 파일 등에 이와 관련된 내용을 명시해주는 것이 좋겠습니다.

모든 도메인 객체가 immutable하다는 것을 인지하지 못한다면,
Payment 객체의 complete 함수를 호출하고 기존 객체가 바뀌었다고 생각해서 기존 객체를 사용해버리는 실수가 있을 수 있을 것 같아요.

song-younghoon commented on src/main/java/kr/hhplus/be/server/domain/concert/Concert.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Objects;
+
+// 콘서트(공연) 도메인 모델
+public final class Concert {
```
기본적으로 모든 도메인 객체에 의존성이 없는 것을 확인하였습니다.
이것이 클린 아키텍처의 핵심이자, SOLID 원칙을 잘 준수하는 프로젝트를 구성하는 시작이라고 보시면 될 것 같습니다.

song-younghoon commented on src/main/java/kr/hhplus/be/server/application/usecase/concert/ConcertService.java:
```diff
@@ -0,0 +1,50 @@
+
+    private final ConcertRepository concerts;              // 도메인 포트
+    private final ConcertScheduleJpaRepository schedules;
+    private final ConfirmedReservationJpaRepository confirmed;// 확정 좌석 조회
```
유즈케이스(in port interface) 구현체가 JPA 객체에 직접 의존하고 있죠?
이러면 DIP를 위배하게 됩니다. 나중에 영속성 레이어의 구현체가 JPA가 아닌 다른 것으로 교체된다면 여기까지 손대야하는 문제가 생깁니다. interface에 의존하고 주입받도록 교정하는게 좋겠습니다.

song-younghoon commented on src/main/java/kr/hhplus/be/server/application/usecase/payment/Wallet.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Objects;
+import java.util.UUID;
+
+public class Wallet {
```
usecase 디렉토리에는 in port에 대한 구현체를 담고 있는 Application Layer라고 생각했는데, 여기에 도메인 객체가 정의되어있네요!
도메인 객체들에 대한 정의는 domain 디렉토리 안에 넣어두는게 좋겠습니다.

클래스가 정의된 위치가 일관성이 없으면 프로젝트의 전반적인 구조를 파악하는데 상당한 어려움이 따르게됩니다 😢

song-younghoon commented on src/main/java/kr/hhplus/be/server/application/port/out/ConcertPort.java:
```diff
@@ -0,0 +1,8 @@
+import kr.hhplus.be.server.infrastructure.persistence.concert.jpa.entity.ConcertJpaEntity;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ConcertPort extends JpaRepository<ConcertJpaEntity, Long> {
```
여기도 out port 인터페이스가 JpaRepository를 상속하면 안되죠!
JpaRepository는 어뎁터의 형태로 out port를 구현하여 애플리케이션 레이어(Service)에서 사용될때 인터페이스에 주입되어야합니다!

song-younghoon reviewed (approved):
우선 전반적으로 강의에서 배운 내용을 잘 적용하려고 노력한 티가 납니다.
또한 이미 코멘트 드린 사항이지만 아키텍처와 무관하게, 도메인 객체를 immutable하게 다뤄주셨다는 점이 너무 훌륭해서 한번 더 훌륭하다는 말씀 드립니다 ㅎㅎ

다만 현재 배웠던 개념들이 명확하게 정리가 잘 안되어있는것 같다는 느낌을 받았어요.
프로젝트의 전반적인 구조는 전형적인 헥사고날 아키텍처의 패턴을 따르고 있는데, 서비스에서 직접 JPA 인터페이스에 의존한다거나, usecase에 도메인 객체가 정의되어있는 등 전반적으로 프로젝트가 정리되지 않아 구조를 파악하는데 조금 힘들었습니다 😢 

물론 디렉토리 구조가 핵심인것은 아닙니다만, 어느정도 일관된 구조를 가지고 있어야 프로젝트가 전반적으로 어떻게 구성되어있는지 파악하기 수월하다는 사실을 알고 조금 정리가 되면 좋을 것 같아요!

또한 레이어드 아키텍처든, 클린 아키텍처든, 헥사고날 아키텍처든 .. 저는 개인적으로 모든 아키텍처 패턴들이 가장 강하게 강조하고 있는 것은 DIP(의존성 역전 원칙)이라고 생각합니다.
서비스는 범용적인 인터페이스에 의존해야해요. 특정 기술에 종속되는 인터페이스나 구현체에 의존하면 안됩니다.

따라서 레이어드 아키텍처도 사실 서비스에서 JpaRepository가 아니라 그냥 Repository 인터페이스에 의존하고, 실제 런타임에서는 스프링이 이 Repository 인터페이스에 JPA 구현체를 꽂아넣도록 구성해야해요.
(우리가 스프링과 같은 프레임워크를 사용해서 얻을 수 있는 가장 큰 이점 중에 하나입니다. 스프링의 IoC 덕분에 인터페이스에 구현체를 꽂아넣는 작업을 손쉽게 세팅할 수 있죠 ㅎㅎ)

주요 사항을 정리하자면 다음과 같습니다!
- port/in 에는 컨트롤러가 호출하는 메소드의 인터페이스를 정의하도록 합시다. (이부분은 잘 하셨어요!)
- port/out 에는 서비스에서 의존이 필요한 인터페이스들이 여기에 다 몰려있도록 합시다.
  - 예를 들어 QueuePort와 QueueRepository는 동일한 역할을 수행하는 인터페이스로 보이는데, 중복 정의 되어있어서 무슨 차이인지 알기가 어렵습니다...!
- dto는 컨트롤러에 대한 dto라면, web 디렉토리 안에 넣어버립시다.
- query는 정확히 무슨 용도로 존재하는 디렉토리인지 이해하기 어려운데, usecase의 구현체(서비스)라면 usecase 디렉토리 안에 넣어버립시다!
- 도메인 객체는 전부 domain 디렉토리에 몰아넣어주세요!

---

이어서 리뷰 받고 싶은 내용에 대한 답변 드리겠습니다!

## 테스트 시 Mocking이 적절한가?
> 예약/결제 기능을 UseCase 단위로 나누고 외부 의존성을 Mock으로 대체했는데, 지금처럼 SeatHoldPort, ReservationRepository 등을 다 Mock 처리하는 방식이 과연 적절한지 궁금합니다. 혹시 제가 너무 단순하게 분리한 건 아닌지, 충분한지 궁금합니다.

적절하게 잘 수행해주셨습니다.
원래 서비스에서는 인터페이스에 의존하기 때문에 구현체를 끌어오는것 자체가 쉽지 않고, 유닛 테스트에서는 원칙적으로 의존하는 객체는 mocking 하는게 맞습니다.
유닛 테스트를 작성하려니까 테스트 대상이 되는 클래스가 전부 인터페이스에 의존하고 있어서 자연스럽게 mocking하게 되죠?

클래스를 순수하고 testable하게 잘 작성했다는 의미입니다.

## 도메인과 JPA 엔티티 분리 필요성
> Wallet(도메인)과 UserWallet(JPA 엔티티)을 분리했는데, 실제로는 역할이 많이 겹치는 것 같아 중복 코드처럼 보였습니다. 과제 수준에서는 그냥 합쳐도 됐을까요? 아니면 장기적으로는 지금처럼 분리하는 게 좋은 선택일까요?

실제로 코드가 거의 겹칠지라도 클린 아키텍처 관점에서는 분리시키는게 맞습니다.
왜냐면 out port(Repository) 인터페이스 입장에서는 JPA 엔티티는 전혀 관심사가 아니기 때문이죠.
JPA 엔티티는 단순히 Repository를 구현하는 구현체(JPA)에서 필요한 객체입니다.
즉, Wallet과 UserWallet(사실 WalletJpaEntity라고 명명하는게 더 맞을 것 같긴합니다만 ...)은 전혀 다른 객체입니다.

그리고 UserWallet은 무슨일이 있어도 JPA 구현체 밖으로 나오면 안돼요.
즉, out port와 service는 UserWallet이라는 객체의 존재를 몰라야합니다. (당연히 in port도 😄 )

잘 이해가 안되신다면 이렇게 생각해보셔도 좋아요.
JPA 구현체에서 Wallet 객체를 반환하기 위해서 내부적으로는 여러개의 테이블을 참조할 수도 있잖아요? 그런 경우에는 Wallet을 구성하기 위한 여러개의 JpaEntity가 존재할 수도 있는 것입니다.

## 중복과 구조적인 실수
> 구현하면서 WalletService와 WalletJpaAdapter가 동시에 존재하는 등 중복된 구조가 생겼다가 뒤늦게 정리했습니다. 이런 상황을 피하려면 처음 설계할 때 어떤 기준으로 “여긴 Adapter”, “여긴 Service”라고 선을 긋는 게 좋을까요?

선을 긋는 기준은 아주 명확합니다.

Adapter는 Repository의 구현체이구요. Service는 도메인 객체를 다뤄서 비즈니스 로직을 실행시키는 곳입니다.
따라서 단순히 잔액을 확인하는 메소드를 호출할 때도 아래와 같은 흐름을 타는게 맞습니다.
- controller에서 '잔액 조회' 메소드가 호출되면 in port에 정의된 인터페이스를 호출 (실제 런타임에서는 스프링이 이 인터페이스의 구현체인 service를 주입시킴)
- service는 in port를 implement 해야함
- service는 out port의 인터페이스에 의존해서 잔액 조회 메소드를 호출해야함 (예를 들어 WalletPort)
  - 실제 런타임에서는 스프링이 이 WalletPort에 해당 인터페이스의 구현체인 JpaRepository를 주입시킬 것임
- WalletPort를 implement하는 JpaRepository를 어뎁터쪽에 구현

이렇게 되면 나중에 WalletPort의 구현체가 바뀌더라도 인터페이스는 그대로이기 때문에 서비스를 건들 필요가 없습니다. (Jpa를 버리고 MongoDB로 교체해야하는 상황에서 WalletPort를 implement하는 WalletMongoRepository만 새로 구현하고 스프링 Bean으로 등록하면 끝.)

## 레이어드 vs 클린 아키텍처 경계
> 이번 과제에서는 Concert 조회/Wallet 충전은 레이어드로, 예약/결제는 클린 아키텍처로 나눴는데요, 실제 현업에서도 이렇게 혼합해서 쓰는 경우가 많은지, 아니면 보통 일관되게 하나로 가져가는지 궁금합니다.

코멘트 드렸다시피 약간 현재 프로젝트의 형태는 헥사고날에 가깝긴 한데요,
보통 처음부터 헥사고날이나 클린 아키텍처로 구조를 잡아간다면 그걸로 일관되게 쭉 가게 되구요!
이미 만들어진 프로젝트를 유지보수하는 경우에는 해당 프로젝트가 레이어드 아키텍처로 구성되어있는 경우가 흔하기 때문에 여기서 더 OOP를 잘 활용하게 프로젝트를 개선해보자! 하면 기존의 프로젝트 구조를 거의 유지하면서 클린 아키텍처의 핵심 철학을 조금씩 반영하는 형태로 리팩토링합니다.

말씀 드렸다시피, 디렉토리 구조가 중요한건 아니고 헥사고날/클린 아키텍처의 핵심은 '추상화된 인터페이스에 의존한다, 도메인 객체는 순수해야한다.(-> 이걸 되게 뭉뚱그려서 '안쪽은 바깥쪽을 모른다.'라고 표현하긴 하지만 제가 보기에는 이 말이 더 이해하기 어려운 말 같아요.)'입니다.

이에 맞춰서 서비스 레이어에서 DB의 엔티티를 바로 도메인 객체처럼 다루고 있다면, 중간에 도메인 객체 레이어를 하나 더 두는 식으로 조금씩 바꿔나가는거죠.

## Service vs Adapter 경계와 네이밍이 충분히 읽히나요?
> 초기에 책임이 섞였던 걸 *UseCase / *Port / *JpaAdapter / *JpaRepository로 정리했습니다. 지금 구조/네이밍만으로도 역할이 명확히 보이는지, 더 나을 수 있는 기준이 있다면 코멘트 부탁드립니다.

뭐든 일관성이 중요합니다. (개인 프로젝트라면 프로젝트 내에서의 일관성, 팀이라면 팀 내에서의 모든 프로젝트에서 일관성)
이 일관성만 잘 지켜진다면 Postfix가 무엇이 되었던 그게 크게 중요하지는 않습니다.

처음에는 디렉토리 안에 막 여러 클래스들이 혼재되어있는 케이스가 있어서 파악에 시간이 좀 걸리긴 했지만,
궁극적으로 말씀주신 구조로 프로젝트를 빌드했다는 사실을 인지하는데 클래스 네이밍때문에 크게 어려움을 겪지는 않았습니다.

저는 다음과 같이 이해했고, 이게 예진님이 의도하신 바가 맞다면 문제 없습니다.
- *UseCase: in port 인터페이스
- *Port: out port 인터페이스
- *Service: in port 인터페이스 구현체 (비즈니스 로직을 담고 있는 중앙 레이어)
- *JpaAdapter: out port 인터페이스 구현체
- *JpaRepository: *JpaAdapter 구현을 위해 필요한 JPA 레포지토리

---

벌써 STEP 3가 끝나고 STEP 4로 넘어갈차례네요 🎉 
이번에도 정말 고생 많으셨고, 코멘트 드린 내용 반영하여 더 나은 구조로 개선해주시면 좋겠습니다!

항상 말씀 드리는거지만 처음부터 완벽한건 없어요.
구조도 그렇고 테스트도 그렇고. 시나리오를 진행하는 동안에 계속 다듬고, 여기에 새로 배우는 테크닉들을 하나하나 추가해나가다 보면 어느새 정말 많은것들이 고려되어있는 멋진 프로젝트가 되어있을겁니다!
파이팅 🔥

moonyaejin committed:
refactor: 헥사고날 아키텍처에서 클린 아키텍처로 전면 마이그레이션

- 클린 아키텍처 원칙에 따라 프로젝트 구조 재구성
- 도메인, 애플리케이션, 인프라, 웹 계층 완전 분리
- 의존성 역전을 위한 Port & Adapter 패턴 구현
- 도메인 서비스를 애플리케이션 서비스에서 분리
- 도메인 계층의 인프라 의존성 완전 제거

세부 개선사항:
- port/in: 컨트롤러가 호출하는 UseCase 인터페이스만 정의
- port/out: 서비스에서 필요한 모든 외부 의존성 인터페이스 통합
- 중복 인터페이스 제거
- DTO를 web 패키지로 이동하여 계층 책임 명확화
- 모든 도메인 객체를 domain 패키지로 통합

moonyaejin committed:
refactor: Repository 인터페이스를 Port로 일원화

moonyaejin committed:
refactor: 아키텍처 분리 - 예약/결제는 클린, 조회/충전은 레이어드로 구현

moonyaejin committed:
feat: 기존 토큰 시스템을 실제 대기열로 개선

moonyaejin committed:
test: WalletService 단위 테스트 추가 및 클린 아키텍처 개선

song-younghoon commented:
열심히 개선하고 계시는군요 👍 
다 지켜보고 있습니다 너무 훌륭해요 😄 파이팅 🔥
